contracts/evm/src/eth/utils/SparseMerkleTree.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.12;
+
+/**
+ * @title Sparse Merkle Tree library
+ * @notice Implements proof verification for SMTs, which is meant to be used
+ * when dealing with the checkpoint SMTs
+ * @dev This implementation is based on https://github.com/pokt-network/smt.
+ * It uses zero as the nil node value and implements similar optimizations to
+ * those discussed in https://ethresear.ch/t/optimizing-sparse-merkle-trees/3751/4
+ * and related articles. Proofs used here can be referred to as 'compact'
+ * proofs in some client implementations because unnecessary side nodes are
+ * omitted and indicated through a bitmask.
+ * This is not a fixed-size SMT - its maximum depth is 256. This means proofs
+ * side nodes are not always 256 in its decompacted form, though it also means
+ * inner nodes and leaves hash functions are different - leaves are hashed
+ * with `keccak256(abi.encodePacked(uint8(0), path, value))` and inner nodes
+ * are hashed with `keccak256(abi.encodePacked(uint8(0), left, right))`
+ */
+library SparseMerkleTree {
+    struct Proof {
+        bytes32 key;
+        bytes32 value;
+        uint256 bitMask;
+        bytes32[] sideNodes;
+        uint256 numSideNodes;
+        bytes32 nonMembershipLeafPath;
+        bytes32 nonMembershipLeafValue;
+    }
+
+    /**
+     * @notice Maximum SMT depth, which is also the maximum proof side nodes
+     * length
+     */
+    uint256 internal constant MAX_DEPTH = 256;
+
+    /**
+     * @notice Nil value in the SMT
+     */
+    bytes32 internal constant DEFAULT_VALUE = bytes32(0);
+
+    /**
+     * @notice Leaf node prefix for hashing
+     */
+    uint8 internal constant LEAF_NODE_PREFIX = 0;
+
+    /**
+     * @notice Inner node prefix for hashing
+     */
+    uint8 internal constant INNER_NODE_PREFIX = 1;
+
+    /**
+     * @notice Verifies an SMT (non-)inclusion proof, checking that the
+     * resulting root is the same as the current root
+     * @param root Current SMT root
+     * @param proof SMT proof
+     * @return Whether the (non-)inclusion proof is valid
+     */
+    function verifyProof(bytes32 root, Proof calldata proof) internal pure returns (bool) {
+        require(proof.sideNodes.length <= MAX_DEPTH && proof.numSideNodes <= MAX_DEPTH, "Side nodes exceed depth");
+
+        bytes32 path = keccak256(abi.encodePacked(proof.key));
+        bytes32[3] memory hashBuffer;
+
+        bytes32 currentHash;
+
+        if (proof.value == DEFAULT_VALUE) {
```

Comment:
I thought I'd still agree with you on that, but started trying it out and honestly. besides gas considerations (minimal), I feel like it's not enough nesting to justify the encapsulation, and it's like 10 lines atm. One more place to read in order to understand what's happening, and this is only used in one situation to determine the leaf we're trying to prove. I'd still be up to keeping it as it is.

Reply:
This whole logic just resolves current hash. It's a separate logic unit which is part of algorithm you implemented which just complicates overall already not easy to read algorithm. 

Imo resulting code [here](https://github.com/NethermindEth/near-sffl/pull/53#discussion_r1548071665 ) looks much more readable. And if one decides to check how actually currentHash computed he can just navigate to the function.

Reply:
As one who actually had to read without a lot of context on algorithm itself, initial version of this math trickery was a bit of challange to understand. Too much things going on: magical bitshifts, magical offsets, if elses on top.

Reply:
Let me try again and see how it looks. Usually MT verifiers have similar implementations, so that's maybe why it doesn't seem strange to me. In the end, what happens in this part is simply computing the leaf it's trying to prove based on `value` and `nonMembershipLeaf*`, which is the 'initial hash' before going to the inner nodes.

Reply:
Done in [8e108e2](https://github.com/NethermindEth/near-sffl/pull/53/commits/8e108e24a2d9b1aabba48f652c0986771164130f). Also added `_computeProofRoot` on [8b822e6](https://github.com/NethermindEth/near-sffl/pull/53/commits/8b822e6a28e35d0ac64b477d570e65a76d8f263f).